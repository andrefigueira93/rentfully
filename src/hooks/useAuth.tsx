import { SignUpProps } from "@/schemas/auth";
import { ListUserProps } from "@/schemas/user";
import { authUserAtom, registeredUsersAtom } from "@/store/auth-store";
import bcrypt from "bcryptjs";
import { useAtom } from "jotai";
import React, {
  PropsWithChildren,
  createContext,
  useContext,
  useMemo,
} from "react";

// Define the shape of the context
interface AuthContextData {
  user: Partial<ListUserProps[number]> | null;
  isAuthenticated: boolean;
  signUp: ({ user }: { user: SignUpProps }) => void;
  signIn: ({ email, password }: { email: string; password: string }) => void;
  signOut: () => void;
}
// Create a context with the default value of undefined
const AuthContext = createContext<AuthContextData | undefined>(undefined);

// Create a provider component that will expose the context
const AuthProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [user, setUser] = useAtom(authUserAtom);
  const [registeredUsers, setRegisteredUsers] = useAtom(registeredUsersAtom);

  // Memoed Check if the user is authenticated
  const isAuthenticated = useMemo(() => !!user?.id, [user]);

  // Function to find a user by email
  function findUserByEmail(email: string) {
    return registeredUsers.find((item) => item.email === email);
  }

  // Sign up function to create a new user
  function signUp({ user }: { user: SignUpProps }) {
    const userExists = findUserByEmail(user.email);

    if (userExists?.id) {
      throw new Error("User already exists");
    }

    let newUser = {};

    setRegisteredUsers((prev) => {
      newUser = {
        ...user,
        id: prev.length + 1,
        createdAt: new Date(),
      };
      return [...prev, newUser];
    });

    setUser(newUser);
  }

  // Sign in function to authenticate a user
  function signIn({ email, password }: { email: string; password: string }) {
    const user = findUserByEmail(email);

    if (!user?.id) {
      throw new Error("User not found");
    }

    // Compare the password with the encrypted password
    const decryptedPassword = bcrypt.compareSync(
      password,
      user?.password || ""
    );

    if (!decryptedPassword) {
      throw new Error("Invalid credentials");
    }

    delete user.password;

    setUser(user);
  }

  function signOut() {
    setUser(null);
  }

  return (
    <AuthContext.Provider
      value={{ user, isAuthenticated, signIn, signUp, signOut }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;

export const useAuth = () => {
  const ctx = useContext(AuthContext);
  if (!ctx) throw new Error("`useAuth` must be used within a AuthProvider");
  return ctx;
};
