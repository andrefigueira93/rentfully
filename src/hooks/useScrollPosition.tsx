import React, {
  PropsWithChildren,
  createContext,
  useContext,
  useEffect,
} from "react";

// Define the shape of the context
interface ScrollPositionContextData {
  scrollPosition: {
    x: number;
    y: number;
  };
}

// Create a context with the default value of undefined
const ScrollPositionContext = createContext<
  ScrollPositionContextData | undefined
>(undefined);

// Create a provider component that will expose the context
const ScrollPositionProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [scrollPosition, setScrollPosition] = React.useState({
    x: 0,
    y: 0,
  });

  useEffect(
    () => {
      const handleScroll = () => {
        setScrollPosition({ x: window.scrollX, y: window.scrollY });
      };

      // Add the event listener
      window.addEventListener("scroll", handleScroll);

      // Remove the event listener on cleanup
      return () => {
        window.removeEventListener("scroll", handleScroll);
      };
    },
    [
      // We don't want to run this effect on every render
      // So I'm gonna left this array of dependencies empty
      // https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects
    ]
  );

  return (
    <ScrollPositionContext.Provider value={{ scrollPosition }}>
      {children}
    </ScrollPositionContext.Provider>
  );
};

export default ScrollPositionProvider;

// Create a hook to use the ScrollPositionContext
export const useScrollPosition = () => {
  const ctx = useContext(ScrollPositionContext);
  if (!ctx)
    throw new Error(
      "`useScrollPosition` must be used within a ScrollPositionProvider"
    );
  return ctx;
};
