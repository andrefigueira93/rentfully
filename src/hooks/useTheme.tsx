import { themeAtom } from "@/store/global-store";
import { useAtom } from "jotai";
import { createContext, useContext, useEffect } from "react";

// Theme types
type Theme = "dark" | "light" | "system";

// Define the shape of the props
type ThemeProviderProps = {
  children: React.ReactNode;
};

// Define the shape of the context
type ThemeProviderState = {
  theme: Theme;
  setTheme: (theme: Theme) => void;
};

// Define the initial state of the context
const initialState: ThemeProviderState = {
  theme: "system",
  setTheme: () => null,
};

// Create a context with the default value of undefined
const ThemeProviderContext = createContext<ThemeProviderState>(initialState);

// Create a provider component that will expose the context
export function ThemeProvider({ children, ...props }: ThemeProviderProps) {
  const [theme, setTheme] = useAtom(themeAtom);

  // Set the theme on first load
  useEffect(() => {
    const root = window.document.documentElement;

    root.classList.remove("light", "dark");

    if (theme === "system") {
      const systemTheme = window.matchMedia("(prefers-color-scheme: dark)")
        .matches
        ? "dark"
        : "light";

      root.classList.add(systemTheme);
      return;
    }

    root.classList.add(theme);
  }, [theme]);

  const value = {
    theme,
    setTheme: (theme: Theme) => {
      setTheme(theme);
    },
  };

  return (
    <ThemeProviderContext.Provider {...props} value={value}>
      {children}
    </ThemeProviderContext.Provider>
  );
}

// Create a hook to use the ThemeContext
export const useTheme = () => {
  const context = useContext(ThemeProviderContext);

  if (context === undefined)
    throw new Error("useTheme must be used within a ThemeProvider");

  return context;
};
