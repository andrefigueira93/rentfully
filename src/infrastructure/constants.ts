export const AUTH_TOKEN_NAME = "rentfully:token";
export const AUTH_REGISTERED_USERS = "rentfully:registeredUsers";
