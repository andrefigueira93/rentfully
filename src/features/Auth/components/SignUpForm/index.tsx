import { Button } from "@/components/ui/button";
import { Checkbox } from "@/components/ui/checkbox";
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import { useAuth } from "@/hooks/useAuth";
import { SignUpProps, SignUpSchema } from "@/schemas/auth";
import { zodResolver } from "@hookform/resolvers/zod";
import { ExclamationTriangleIcon } from "@radix-ui/react-icons";
import { useMutation } from "@tanstack/react-query";
import bcrypt from "bcryptjs";
import React from "react";
import { useForm } from "react-hook-form";
import { toast } from "sonner";

const salt = bcrypt.genSaltSync(10);

const SignUpForm: React.FC = () => {
  const form = useForm<SignUpProps>({
    resolver: zodResolver(SignUpSchema),
    defaultValues: {
      firstName: "",
      lastName: "",
      email: "",
      password: "",
      terms: false,
    },
  });

  const { signUp } = useAuth();

  const { mutate, isPending } = useMutation<SignUpProps>({
    mutationFn: async () => {
      const formValues = form.getValues();
      const hashedPassword = bcrypt.hashSync(formValues.password, salt);

      // Form values with hashed password
      const values = {
        ...formValues,
        password: hashedPassword,
      };

      // Call API
      // setTimeout(() => {
      signUp({ user: values });
      // }, 1000);
      return values;
    },
    onSuccess: () => {
      toast("Welcome!", {
        description: "Your account has been created",
        icon: <ExclamationTriangleIcon />,
        cancel: {
          label: "Close",
        },
      });

      setTimeout(() => {
        // refresh the page
        // after 2 seconds
        // window.location.reload();
      }, 2000);
    },
    onError: (error) => {
      console.error(error);
      toast("Whoops!", {
        description: "Verify your credentials",
        icon: <ExclamationTriangleIcon />,
        cancel: {
          label: "Close",
        },
      });
    },
  });

  return (
    <Form {...form}>
      <form
        onSubmit={form.handleSubmit(() => {
          mutate();
        })}
        className="space-y-8"
      >
        <FormField
          control={form.control}
          name="firstName"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Your firstname</FormLabel>
              <FormControl>
                <Input
                  testId={`${field.name}-input`}
                  placeholder="Jhon"
                  {...field}
                />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="lastName"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Your lastname</FormLabel>
              <FormControl>
                <Input
                  testId={`${field.name}-input`}
                  placeholder="Doe"
                  {...field}
                />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="email"
          render={({ field }) => (
            <FormItem>
              <FormLabel>E-mail</FormLabel>
              <FormControl>
                <Input
                  testId={`${field.name}-input`}
                  placeholder="E-mail"
                  {...field}
                />
              </FormControl>
              <FormDescription>Your best email</FormDescription>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="password"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Password</FormLabel>
              <FormControl>
                <Input
                  testId={`${field.name}-input`}
                  placeholder="Password"
                  type="password"
                  {...field}
                />
              </FormControl>
              <FormDescription>Your best password</FormDescription>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="terms"
          render={({ field }) => (
            <FormItem className="flex flex-row items-start space-x-3 space-y-0 rounded-md border p-4 shadow">
              <FormControl>
                <Checkbox
                  checked={field.value}
                  onCheckedChange={field.onChange}
                />
              </FormControl>
              <FormLabel>
                I agree with the Terms and Conditions and the Privacy Policy
              </FormLabel>
            </FormItem>
          )}
        />
        <Button
          data-testid="sign-up-button"
          disabled={isPending}
          role="button"
          type="submit"
          className="mt-2 w-full"
        >
          {isPending ? "Loading..." : "Create Account"}
        </Button>
      </form>
    </Form>
  );
};

export default SignUpForm;
