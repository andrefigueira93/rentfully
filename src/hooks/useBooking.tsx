import { BookingProps, ListBookingProps } from "@/schemas/booking";
import { bookingFiltersAtom } from "@/store/booking-filters-store";
import { authUserBookingsAtom, bookingsAtom } from "@/store/bookings-store";
import { addDays } from "date-fns";
import { useAtom, useAtomValue } from "jotai";
import React, { PropsWithChildren, createContext, useContext } from "react";
import { useNavigate } from "react-router-dom";
import { toast } from "sonner";
import { useAuth } from "./useAuth";

// Define the shape of the context
interface BookingContextData {
  bookings: Partial<ListBookingProps>;
  myBookings: Partial<ListBookingProps>;
  newBooking: (propertyId: number) => void;
  updateBooking: ({ booking }: { booking: Partial<BookingProps> }) => void;
  deleteBooking: (id: number) => void;
}
// Create a context with the default value of undefined
const BookingContext = createContext<BookingContextData | undefined>(undefined);

// Create a provider component that will expose the context
const BookingProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [bookings, setBookings] = useAtom(bookingsAtom);
  const myBookings = useAtomValue(authUserBookingsAtom);
  const filters = useAtomValue(bookingFiltersAtom);
  const { user } = useAuth();

  const navigate = useNavigate();

  // While we don't have a backend, we will use this function to create a new booking
  // and add it to the bookings array with a fake id and createdAt date
  function newBooking(propertyId: number) {
    if (!user?.id) {
      toast("You should be authenticated to make a booking", {
        action: {
          label: "Sign in",
          onClick: () => navigate("/sign-in"),
        },
        cancel: {
          label: "Nevermind",
          onClick: () => toast.dismiss(),
        },
        icon: <span className="text-2xl">ðŸ˜…</span>,
        important: true,
        position: "top-right",
      });

      if (typeof window !== "undefined") {
        throw new Error("You should be authenticated to make a booking");
      }

      return;
    }

    const today = new Date();

    setBookings((prev) => [
      ...prev,
      {
        id: prev.length + 1,
        propertyId,
        userId: user.id!, // We know that user is not undefined because we checked it above
        status: "Pending",
        checkIn: filters.checkIn ?? today, // If checkIn is not defined, use today
        checkOut: filters.checkOut ?? addDays(today, 1), // If checkOut is not defined, use tomorrow
        guests: filters.numberOfGuests,
        createdAt: today,
      },
    ]);

    navigate("/new-booking");
  }

  // Function to update a booking with updatedAt date
  function updateBooking({ booking }: { booking: Partial<BookingProps> }) {
    setBookings((prev) => {
      return prev.map((item) => {
        if (item.id === booking.id) {
          return {
            ...item,
            ...booking,
            updatedAt: new Date(),
          };
        }
        return item;
      });
    });
  }

  // Function to delete a booking
  function deleteBooking(id: number) {
    setBookings((prev) => prev.filter((item) => item.id !== id));
  }

  return (
    <BookingContext.Provider
      value={{
        bookings,
        myBookings,
        newBooking,
        deleteBooking,
        updateBooking,
      }}
    >
      {children}
    </BookingContext.Provider>
  );
};

export default BookingProvider;

// Create a hook to use the BookingContext
export const useBooking = () => {
  const ctx = useContext(BookingContext);
  if (!ctx)
    throw new Error("`useBooking` must be used within a BookingProvider");
  return ctx;
};
