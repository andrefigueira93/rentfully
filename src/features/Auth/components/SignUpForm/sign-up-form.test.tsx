/* eslint-disable @typescript-eslint/no-explicit-any */
import { newTestUser } from "@/__tests__/test-constants";
import TestWrapper from "@/__tests__/wrapper";
import * as AuthHook from "@/hooks/useAuth";
import { AUTH_REGISTERED_USERS } from "@/infrastructure/constants";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";
import SignUpForm from "./index";

vi.mock("bcryptjs");
vi.mock("sonner");
vi.mock("@/hooks/useAuth", async () => {
  const actualModule = await vi.importActual("@/hooks/useAuth");
  return {
    ...(actualModule as any),
    useAuth: () => ({
      signUp: vi.fn(),
    }),
  };
});
vi.mock("@tanstack/react-query", async () => {
  const actualModule = await vi.importActual("@tanstack/react-query");
  return {
    ...(actualModule as any),
    useMutation: () => ({
      mutate: vi.fn(),
      isPending: false,
    }),
  };
});

describe("SignUpForm", () => {
  let container: HTMLElement;
  let firstNameInput: HTMLElement;
  let lastNameInput: HTMLElement;
  let emailInput: HTMLElement;
  let passwordInput: HTMLElement;
  let checkbox: HTMLElement;

  beforeEach(() => {
    vi.clearAllMocks();
    const { container: component } = render(<SignUpForm />, {
      wrapper: TestWrapper,
    });
    container = component;
    firstNameInput = screen.getByTestId("firstName-input");
    lastNameInput = screen.getByTestId("lastName-input");
    emailInput = screen.getByTestId("email-input");
    passwordInput = screen.getByTestId("password-input");
    checkbox = screen.getByRole("checkbox");
  });

  test("renders form fields", () => {
    expect(firstNameInput).toBeInTheDocument();
    expect(lastNameInput).toBeInTheDocument();
    expect(emailInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(checkbox).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("handles form submission with correct credentials", async () => {
    localStorage.removeItem(AUTH_REGISTERED_USERS);

    const mockSignUp = vi.fn();
    vi.spyOn(AuthHook, "useAuth").mockReturnValue({
      user: null,
      isAuthenticated: false,
      signIn: vi.fn(),
      signOut: vi.fn(),
      signUp: mockSignUp,
    });

    fireEvent.change(emailInput, {
      target: { value: newTestUser.email },
    });
    fireEvent.change(passwordInput, {
      target: { value: newTestUser.password },
    });
    fireEvent.change(firstNameInput, {
      target: { value: newTestUser.firstName },
    });
    fireEvent.change(lastNameInput, {
      target: { value: newTestUser.lastName },
    });
    fireEvent.click(checkbox);

    fireEvent.click(screen.getByTestId("sign-up-button"));

    // TODO: Finish this test
    // await waitFor(
    //   () => {
    //     expect(mockSignUp).toHaveBeenCalled();
    //   },
    //   {
    //     timeout: 2000,
    //   }
    // );
  });

  test("shows error on invalid submit", async () => {
    fireEvent.click(screen.getByTestId("sign-up-button"));
    await waitFor(() => {
      expect(screen.getAllByTestId("form-message")).toHaveLength(2);
    });
  });
});
