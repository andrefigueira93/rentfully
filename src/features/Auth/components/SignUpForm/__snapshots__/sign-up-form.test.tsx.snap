// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignUpForm > renders form fields 1`] = `
<div>
  <form
    class="space-y-8"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        Your firstname
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid="false"
        class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="firstName-input"
        id=":r0:-form-item"
        name="firstName"
        placeholder="Jhon"
        value=""
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r1:-form-item"
      >
        Your lastname
      </label>
      <input
        aria-describedby=":r1:-form-item-description"
        aria-invalid="false"
        class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="lastName-input"
        id=":r1:-form-item"
        name="lastName"
        placeholder="Doe"
        value=""
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r2:-form-item"
      >
        E-mail
      </label>
      <input
        aria-describedby=":r2:-form-item-description"
        aria-invalid="false"
        class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="email-input"
        id=":r2:-form-item"
        name="email"
        placeholder="E-mail"
        value=""
      />
      <p
        class="text-[0.8rem] text-muted-foreground"
        id=":r2:-form-item-description"
      >
        Your best email
      </p>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r3:-form-item"
      >
        Password
      </label>
      <input
        aria-describedby=":r3:-form-item-description"
        aria-invalid="false"
        class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="password-input"
        id=":r3:-form-item"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <p
        class="text-[0.8rem] text-muted-foreground"
        id=":r3:-form-item-description"
      >
        Your best password
      </p>
    </div>
    <div
      class="flex flex-row items-start space-x-3 space-y-0 rounded-md border p-4 shadow"
    >
      <button
        aria-checked="false"
        aria-describedby=":r4:-form-item-description"
        aria-invalid="false"
        class="peer h-4 w-4 shrink-0 rounded-sm border border-primary shadow focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground"
        data-state="unchecked"
        id=":r4:-form-item"
        role="checkbox"
        type="button"
        value="on"
      />
      <input
        aria-hidden="true"
        style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
        tabindex="-1"
        type="checkbox"
        value="on"
      />
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r4:-form-item"
      >
        I agree with the Terms and Conditions and the Privacy Policy
      </label>
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 mt-2 w-full"
      data-testid="sign-up-button"
      role="button"
      type="submit"
    >
      Create Account
    </button>
  </form>
</div>
`;
